nemo/collections/tts/modules/submodules.py

```diff
@@ -410,3 +419,284 @@ def forward(self, forward_input: Tuple[torch.Tensor, torch.Tensor]):
                 output = output + res_skip_acts
 
         return self.end(output)
+
+
+class WeightedSpeakerEmbedding(torch.nn.Module):
+    def __init__(self, pretrained_embedding, speaker_list=[]):
+        super().__init__()
+        if len(speaker_list) > 0:
+            self.pretrained_embedding = torch.nn.Parameter(pretrained_embedding.weight[speaker_list].detach().clone())
+        else:
+            self.pretrained_embedding = torch.nn.Parameter(pretrained_embedding.weight.detach().clone())
+        self.pretrained_embedding.requires_grad = False
+        self.num_embeddings = self.pretrained_embedding.size()[0]
+        self.embedding_weight = torch.nn.Parameter(torch.ones(1, self.num_embeddings))
+
+    def forward(self, speaker):
+        weight = self.embedding_weight.repeat(len(speaker), 1)
+        weight = torch.nn.functional.softmax(weight, dim=-1)
+        speaker_emb = weight @ self.pretrained_embedding
+        return speaker_emb
+
+
+class GlobalStyleToken(NeuralModule):
+    """
+    Global Style Token based Speaker Embedding
+    """
+
+    def __init__(
+        self,
+        cnn_filters=[32, 32, 64, 64, 128, 128],
+        dropout=0.2,
+        gru_hidden=128,
+        gst_size=128,
```

Comment:
Should we pass in the ReferenceEncoder directly to __init__?

Reply:
Yes. I add ReferenceEncoder as a module which can be init in config.